cachi2/core/models/output.py

```diff
@@ -1,80 +1,58 @@
 import string
 from pathlib import Path
-from typing import Annotated, Literal, Optional, Union
+from typing import Any, Literal, Optional
 
 import pydantic
 
-from cachi2.core.models.validators import check_sane_relpath, unique_sorted
+from cachi2.core.models.validators import unique_sorted
 
-# Supported package types (a superset of the supported package *manager* types)
-PackageType = Literal["gomod", "go-package", "pip"]
 
+class Component(pydantic.BaseModel):
+    """A software component such as a dependency or a package.
 
-class _DependencyBase(pydantic.BaseModel):
-    """Common attributes for all dependency types."""
-
-    type: PackageType
-    name: str
-
-
-class GomodDependency(_DependencyBase):
-    """Metadata about a gomod dependency."""
-
-    type: Literal["gomod"]
-    version: str
-
-
-class GoPackageDependency(_DependencyBase):
-    """Metadata about a go-package dependency.
-
-    Unlike other dependency types, go-package dependencies may come from the standard library,
-    in which case their version is null.
+    Compliant to the CycloneDX specification:
+    https://cyclonedx.org/docs/1.4/json/#components
     """
 
-    type: Literal["go-package"]
+    name: str
     version: Optional[str]
 
+    type: Literal["library"] = "library"
 
-class PipDependency(_DependencyBase):
-    """Metadata about a pip dependency.
+    # This will soon be replaced by the purl
+    def key(self) -> str:
+        """Uniquely identifies a package.
 
-    Pip dependencies have a 'dev' value to indicate whether the dependency is used at runtime
-    or at build time (requirements.txt X requirements-build.txt).
-    """
-
-    type: Literal["pip"]
-    version: str
-    dev: bool
-
-
-Dependency = Annotated[
-    Union[GomodDependency, GoPackageDependency, PipDependency],
-    pydantic.Field(discriminator="type"),
-]
+        Used mainly for sorting and deduplication.
+        """
+        return f"{self.name}:{self.version}"
 
+    @classmethod
+    def from_package_dict(cls, package: dict[str, Any]) -> "Component":
+        """Create a Component from a Cachi2 package dictionary.
 
-def _dependency_sorting_key(dep: Dependency) -> tuple[str, bool, str, str]:
-    # should return all the attributes a Dependency (the whole Union) can have
-    return dep.type, getattr(dep, "dev", False), dep.name, dep.version or ""
+        A Cachi2 package has extra fields which are unnecessary and can cause validation errors.
+        """
+        return Component(name=package.get("name", None), version=package.get("version", None))
```

Comment:
Maybe versions should only be semver versions? Or at least exclude file paths? Maybe that could help with the deduplication problem :thinking: 

To do that, you would probably want package manager code to return sensible Components rather than taking the dicts (which hardly even relate to the output cachi2 is now meant to produce)

Reply:
We would report the file/repo paths only in the purls, then?

Reply:
Yeah, that's the idea. IMO version should always be the semver version (when we have one). For example, yarn puts the resolved version in yarn.lock even for git dependencies, so the component would be something like
```json
{
  "type": "library",
  "name": "cachito-yarn-without-deps",
  "version": "1.0.0",
  "purl": "pkg:npm/cachito-yarn-without-deps?vcs_url=git+https://github.com/cachito-testing/cachito-yarn-without-deps.git@da0a2888aa7aab37fec34c0b36d9e44560d2cf3e"
}
```
For file paths, you can't report them as versions because of the issue mentioned in the comment above. 
```json
{
  "version": "../backend-shared",
  "purl": "pkg:golang/github.com/redhat-appstudio/managed-gitops/backend-shared@vx.y.z"
}
{
  "version": "../../backend-shared",
  "purl": "pkg:golang/github.com/redhat-appstudio/managed-gitops/backend-shared@vx.y.z"
}
```
^ conflict

Reply:
Yeah, that makes sense. I'll make the `Component.version` attribute semver only, then.